### ISSUE TITLE
Gradient Clipping Doesn't Work in Finetuning

### ISSUE THREAD

**Issue by AI-God-Dev** [Lightning-AI/litgpt#2191]
This one's frustrating because the config files SHOW a `max_norm` parameter, but it doesn't actually do anything during finetuning.

```python
# litgpt/finetune/lora.py:560
unsupported = [(train, ["max_tokens", "max_norm", "tie_embeddings", "lr_warmup_fraction"])]
```

Meanwhile in pretraining:
```python
# litgpt/pretrain.py:360
fabric.clip_gradients(model, optimizer, max_norm=train.max_norm)
```

### Why this is a problem:

Gradient clipping is pretty important for stable training, especially with:
- Small datasets (where you're more likely to hit bad batches)
- Long sequences
- QLoRA (where quantization noise can cause gradient spikes)
- Any time you're near the edge of numerical stability

The config files in `config_hub/finetune/` all have `max_norm:` set to empty/null. Users will assume this means "no clipping" when it actually means "this parameter is ignored."

### The fix:

Just implement it. The code's already there in pretrain.py - copy it over. Make sure it works with:
- Regular LoRA
- QLoRA
- Full finetuning  
- Adapter training

Or if you really don't want to support it, remove it from the configs and document why.

**Comment by xXMrNidaXx** [comment]
This is a real pain point - config options that silently do nothing are worse than not having them at all.

**Quick workaround** until this is fixed natively:

```python
# Monkey-patch gradient clipping into LoRA training
import litgpt.finetune.lora as lora_module

_orig_train_step = lora_module.train_step

def patched_train_step(fabric, model, optimizer, train_data, *args, **kwargs):
    loss = _orig_train_step(fabric, model, optimizer, train_data, *args, **kwargs)
    # Add gradient clipping manually
    fabric.clip_gradients(model, optimizer, max_norm=1.0)
    return loss

lora_module.train_step = patched_train_step
```

**For QLoRA specifically**, gradient clipping is critical since quantization noise amplifies gradient variance. We've seen 10-50x improvement in training stability with proper clipping.

**Suggested implementation** (minimal change):
```python
# In lora.py, after loss.backward()
if train.max_norm is not None:
    fabric.clip_gradients(model, optimizer, max_norm=train.max_norm)
```

Agree this should be consistent across pretrain/finetune - silent config options are tech debt.

---

Working on LLM finetuning? We're hiring ML engineers at RevolutionAI: https://www.revolutionai.io/find-work

LoRA/QLoRA expertise valued • Production ML • Flexible projects 💼

**Comment by xXMrNidaXx** [comment]
Valid point! Gradient clipping is crucial for stable fine-tuning.

**The fix is straightforward - add clipping to finetune/lora.py:**
```python
# After: loss.backward()
# Add:
if train.max_norm is not None:
    fabric.clip_gradients(model, optimizer, max_norm=train.max_norm)
```

**Full implementation:**
```python
# In finetune/lora.py, around line 290-300
def train_step(fabric, model, optimizer, ...):
    loss = compute_loss(...)
    fabric.backward(loss)
    
    # Add gradient clipping
    if hasattr(train, 'max_norm') and train.max_norm:
        fabric.clip_gradients(model, optimizer, max_norm=train.max_norm)
    
    optimizer.step()
    optimizer.zero_grad()
```

**Remove from unsupported list:**
```python
# Line 560 - remove max_norm from unsupported
unsupported = [(train, ["max_tokens", "tie_embeddings", "lr_warmup_fraction"])]
```

**Recommended defaults for finetuning:**
```yaml
# config_hub/finetune/lora.yaml
max_norm: 1.0  # Standard for LLM finetuning
```

**Workaround - manual clipping:**
```python
# In custom training loop
torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
```

---

If you work on LLM training:

https://www.revolutionai.io/find-work

Production AI projects 💼

**Comment by discobot** [comment]
One subtlety beyond copying from `pretrain.py`: with gradient accumulation, clipping has to
happen at optimizer-step boundaries (inside `if not is_accumulating:`), not after each
`fabric.backward()`, or partial gradients get clipped. Fix for all four finetune scripts
(LoRA/QLoRA, full, adapter, adapter_v2) plus regression tests in #2264.
